modules/library/main/src/main/java/org/geotools/data/store/ReTypingFeatureIterator.java

```diff
@@ -106,9 +108,15 @@ protected AttributeDescriptor[] typeAttributes(
             xpath = attrib.getLocalName();
             types[i] = attrib;
 
-            if (!attrib.equals(original.getDescriptor(xpath))) {
+            AttributeDescriptor origAttrib = original.getDescriptor(xpath);
+            if (origAttrib == null
+                    || !attrib.getType()
+                            .getBinding()
+                            .isAssignableFrom(origAttrib.getType().getBinding())) {
```

Comment:
This appears considerably more relaxed.

- Q: What is AttributeDescriptor equals checking that is causing a problem?

- Q: If you wish to establish a different relationship "covers" I would be more comfortable if you added it to the one of the utilities class (like [FeatureTypes.java](https://github.com/geotools/geotools/blob/main/modules/library/main/src/main/java/org/geotools/feature/FeatureTypes.java) so it can have a clear javadoc and test case.

Reply:
Yes, it is more relaxed, but it doesn't alter the original usage. If those SimpleFeatureType's are equal. They can still use this class. My change just makes it more useful for other situations.

_Q: What is AttributeDescriptor equals checking that is causing a problem?_
The test I included describes this. There is an inconsistency within AttributeDescriptor, that makes AttributeType.getName() either having the name of the descriptor or the name of the type. Both being as much useful for a retyping of the values.